File: liwc.js
#!/usr/bin/env node
var fs = require('fs');

function extend(array, l) {
  return array.push.apply(array, l);
}

function escapeRegex(text) {
  // from http://simonwillison.net/2006/Jan/20/escape/#p-6
  return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
}

var json_trie = fs.readFileSync('/usr/local/data/liwc_2007.trie', 'utf8'),
  _trie = JSON.parse(json_trie),
  categories = ['funct', 'pronoun', 'ppron', 'i', 'we', 'you', 'shehe', 'they', 'ipron', 'article', 'verb', 'auxverb', 'past', 'present', 'future', 'adverb', 'preps', 'conj', 'negate', 'quant', 'number', 'swear', 'social', 'family', 'friend', 'humans', 'affect', 'posemo', 'negemo', 'anx', 'anger', 'sad', 'cogmech', 'insight', 'cause', 'discrep', 'tentat', 'certain', 'inhib', 'incl', 'excl', 'percept', 'see', 'hear', 'feel', 'bio', 'body', 'health', 'sexual', 'ingest', 'relativ', 'motion', 'space', 'time', 'work', 'achieve', 'leisure', 'home', 'money', 'relig', 'death', 'assent', 'nonfl', 'filler'],
  full_columns = ['WC', 'WPS', 'Sixltr', 'Dic', 'Numerals'],
  punctuations = [
    {name: 'Period', chars: '.'},
    {name: 'Comma', chars: ','},
    {name: 'Colon', chars: ':'},
    {name: 'SemiC', chars: ';'},
    {name: 'QMark', chars: '?'},
    {name: 'Exclam', chars: '!'},
    {name: 'Dash', chars: '-'},
    {name: 'Quote', chars: '"'},
    {name: 'Apostro', chars: "'"},
    {name: 'Parenth', chars: '()[]{}'},
    {name: 'OtherP', chars: '#$%&*+-/<=>@\\^_`|~'}
  ];
extend(full_columns, categories); // exclude filler? is that a bug?
extend(full_columns, ['Period', 'Comma', 'Colon', 'SemiC', 'QMark', 'Exclam', 'Dash', 'Quote', 'Apostro', 'Parenth', 'OtherP', 'AllPct']);
for (var p in punctuations) {
  var raw = punctuations[p].chars.split('').map(escapeRegex).join('|');
  punctuations[p].regex = new RegExp(raw, 'ig');
}

// async -- too much overhead?
// function init(callback) {
  // fs.readFile('/usr/local/data/liwc_2007.trie', 'utf8', function (err, data) {
  //   if (err) throw err;
  //   _trie = JSON.parse(data);
  //   callback(err);
  // });
// }

function _walk(token, i, cursor) {
  if ('*' in cursor)
    return cursor['*'];
  if ('$' in cursor && i === token.length)
    return cursor['$'];
  if (i < token.length) {
    var letter = token[i];
    if (letter in cursor)
      return _walk(token, i + 1, cursor[letter]);
  }
  return [];
}
function from_tokens(tokens) {
  var counts = {Dic: 0, WC: tokens.length};
  for (var i in categories)
    counts[categories[i]] = 0;

  for (var j in tokens) {
    var cats = _walk(tokens[j], 0, _trie);
    if (cats.length) {
      for (var k in cats) {
        counts[cats[k]]++;
      }
      counts.Dic++;
    }
  }

  return counts;
}

function from_text(text, opts) {
  if (opts === undefined) opts = {normalize: true};
  // 'split' produces about 300 more on the brown corpus than 'match' below
  var tokens = text.toLowerCase().split(/[^'a-z0-9A-Z]+/ig);
  // var tokens = text.toLowerCase().match(/[a-z]['a-z]*/g) || [''];
  var sentence_count = (text.match(/[.!?]+/g) || ['']).length;
  var sixltr = 0, numerals = 0;
  for (var i in tokens) {
    var token = tokens[i];
    if (token.length > 6)
      sixltr++;
    if (token.match(/^\d+$/))
      numerals++;
  }

  var counts = from_tokens(tokens);
  // Words per sentence is kind of weird if we're not proceeding incrementally.
  // We just count min(sentence-markers, 1) if normalize == false.
  // But then, the natural thing to do would be to count sentences per word in the LIWC analysis,
  // like everything else, and calculate 1/sentences per word, in post-processing, to get WPS
  counts.WPS = opts.normalize ? counts.WC / sentence_count : sentence_count;
  counts.Sixltr = sixltr;
  counts.Numerals = numerals;

  counts.AllPct = 0;
  for (var j in punctuations) {
    var punc = punctuations[j];
    counts[punc.name] = (text.match(punc.regex) || []).length;
    counts.AllPct += counts[punc.name];
  }
  // N.B.: I am ignoring the LIWC standard of dividing parentheses by two!
  // counts.Parenth = counts.Parenth / 2.0;
  // for (var k in punctuations)

  if (opts.normalize) {
    for (var c = 2; column = full_columns[c]; c++)
      counts[column] = counts[column] / counts.WC;
  }

  return counts;
}

// def print_liwc_results(counts):
//     values = ['%d' % counts['WC'], '%0.2f' % counts['WPS']] + \
//         ['%0.2f' % (counts[column] * 100) for column in full_columns[2:]]
//     for col, val in zip(full_columns, values):
//         print '%16s %s' % (col, val)

module.exports = {
  categories: categories,
  full_columns: full_columns,
  from_tokens: from_tokens,
  from_text: from_text
};

if (require.main === module) {
  var stdin = '';
  process.stdin.on('data', function(chunk) { stdin += chunk; });
  process.stdin.on('end', function() {
    var counts = from_text(stdin);
    console.dir(counts);
  });
  process.stdin.resume();
}

File: lexicons/liwc.py
# -*- coding: utf-8 -*-
import re
from lexicons.base import Lexicon
from collections import Counter
import json


class Liwc(Lexicon):
    corpus_filepath = '/usr/local/data/liwc_2007.trie'

    # category analysis variables:
    category_keys = ['funct', 'pronoun', 'ppron', 'i', 'we', 'you', 'shehe',
        'they', 'ipron', 'article', 'verb', 'auxverb', 'past', 'present', 'future',
        'adverb', 'preps', 'conj', 'negate', 'quant', 'number', 'swear', 'social',
        'family', 'friend', 'humans', 'affect', 'posemo', 'negemo', 'anx', 'anger',
        'sad', 'cogmech', 'insight', 'cause', 'discrep', 'tentat', 'certain',
        'inhib', 'incl', 'excl', 'percept', 'see', 'hear', 'feel', 'bio', 'body',
        'health', 'sexual', 'ingest', 'relativ', 'motion', 'space', 'time', 'work',
        'achieve', 'leisure', 'home', 'money', 'relig', 'death', 'assent', 'nonfl',
        'filler']

    # full analysis variables:
    meta_keys = ['WC', 'WPS', 'Sixltr', 'Dic', 'Numerals']
    puncuation_keys = [
        'Period', 'Comma', 'Colon', 'SemiC', 'QMark', 'Exclam',
        'Dash', 'Quote', 'Apostro', 'Parenth', 'OtherP', 'AllPct']
    punctuation = [
        ('Period', '.'),
        ('Comma', ','),
        ('Colon', ':'),
        ('SemiC', ';'),
        ('QMark', '?'),
        ('Exclam', '!'),
        ('Dash', '-'),  # –—
        ('Quote', '"'),  # “”
        ('Apostro', "'"),  # ‘’
        ('Parenth', '()[]{}'),
        ('OtherP', '#$%&*+-/<=>@\\^_`|~')
    ]

    def __init__(self):
        self.all_keys = self.meta_keys + self.category_keys[:-1] + self.puncuation_keys

        with open(self.corpus_filepath) as corpus_file:
            self._trie = json.load(corpus_file)

    # standard Lexicon functionality:

    def read_token(self, token, token_i=0, trie_cursor=None):
        if trie_cursor is None:
            trie_cursor = self._trie

        if '*' in trie_cursor:
            for category in trie_cursor['*']:
                yield category
        elif '$' in trie_cursor and token_i == len(token):
            for category in trie_cursor['$']:
                yield category
        elif token_i < len(token):
            letter = token[token_i]
            if letter in trie_cursor:
                for category in self.read_token(token, token_i + 1, trie_cursor[letter]):
                    yield category

    def read_document(self, document, token_pattern=r"[a-z]['a-z]*"):
        for match in re.finditer(token_pattern, document.lower()):
            for category in self.read_token(match.group(0)):
                yield category

    # extra (legacy) Liwc functionality:

    def summarize_document(self, document, token_pattern=r"[a-z]['a-z]*", normalize=True):
        sentence_count = len(re.findall(r"[.!?]+", document)) or 1

        # tokens is a bit redundant because it duplicates the tokenizing done
        # in read_document, but to keep read_document simple, we just run it again here.
        tokens = re.findall(token_pattern, document.lower())
        counts = Counter(self.read_document(document, token_pattern=token_pattern))
        counts['Dic'] = sum(counts.values())
        counts['WC'] = len(tokens)
        counts['WPS'] = counts['WC'] / float(sentence_count)
        counts['Sixltr'] = sum(len(token) > 6 for token in tokens)
        counts['Numerals'] = sum(token.isdigit() for token in tokens)

        # count up all characters so that we can get punctuation counts quickly
        character_counts = Counter(document)
        for name, chars in self.punctuation:
            counts[name] = sum(character_counts[char] for char in chars)
        # Parenth is special -- we only count one half of them (to match the official LIWC application)
        counts['Parenth'] = counts['Parenth'] / 2.0
        counts['AllPct'] = sum(counts[name] for name, _ in self.punctuation)

        if normalize:
            # normalize all counts but the first two ('WC' and 'WPS')
            for column in self.all_keys[2:]:
                counts[column] = float(counts[column]) / float(counts['WC'])

        # return a normal dict() rather than the Counter() instance
        result = dict.fromkeys(self.category_keys + ['Dic'], 0)
        result.update(counts)
        return result

    def print_summarization(self, counts):
        absolutes = ['%d' % counts['WC'], '%0.2f' % counts['WPS']]
        percentages = ['%0.2f' % (counts[key] * 100) for key in self.all_keys[2:]]

        for key, value in zip(self.all_keys, absolutes + percentages):
            print '%16s %s' % (key, value)
